actors/builtin/miner/miner_actor.go

```diff
@@ -481,6 +475,158 @@ func (a Actor) SubmitWindowedPoSt(rt Runtime, params *SubmitWindowedPoStParams)
+				return
+			}
+			rt.Log(rtt.INFO, "successfully disputed: %s", err)
+
+			// Ok, now we record faults. This always works because
+			// we don't allow compaction/moving sectors during the
+			// challenge window.
+			//
+			// However, some of these sectors may have been
+			// terminated. That's fine, we'll skip them.
+			faultExpirationEpoch := targetDeadline.Last() + FaultMaxAge
+			powerDelta, err = dlCurrent.RecordFaults(store, sectors, info.SectorSize, QuantSpecForDeadline(targetDeadline), faultExpirationEpoch, disputeInfo.DisputedSectors)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to declare faults")
+
+			err = deadlinesCurrent.UpdateDeadline(store, params.Deadline, dlCurrent)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to update deadline %d", params.Deadline)
+			err = st.SaveDeadlines(store, deadlinesCurrent)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to save deadlines")
+		}
+
+		// Penalties.
+		{
+			// Calculate the base penalty.
+			penaltyBase := PledgePenaltyForInvalidWindowPoSt(
+				epochReward.ThisEpochRewardSmoothed,
+				pwrTotal.QualityAdjPowerSmoothed,
+				disputedPower.QA,
+			)
+
+			// Calculate the target reward.
+			postProofType, err := info.SealProofType.RegisteredWindowPoStProof()
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to determine post proof type")
+			rewardTarget := RewardForDisputedWindowPoSt(postProofType, disputedPower)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to compute reward for disputed window post")
+
+			// Compute the target penalty by adding the
+			// base penalty to the target reward. We don't
+			// take reward out of the penalty as the miner
+			// could end up receiving a substantial
+			// portion of their fee back as a reward.
+			penaltyTarget := big.Add(penaltyBase, rewardTarget)
+
+			err = st.ApplyPenalty(penaltyTarget)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to apply penalty")
+			penaltyFromVesting, penaltyFromBalance, err := st.RepayPartialDebtInPriorityOrder(store, currEpoch, rt.CurrentBalance())
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to pay debt")
+			toBurn = big.Add(penaltyFromVesting, penaltyFromBalance)
+
+			// Now, move as much of the target reward as
+			// we can from the burn to the reward.
+			toReward = big.Min(toBurn, rewardTarget)
+			toBurn = big.Sub(toBurn, toReward)
+
+			pledgeDelta = penaltyFromVesting.Neg()
+		}
+	})
+
+	requestUpdatePower(rt, powerDelta)
+
+	if !toReward.IsZero() {
+		// Try to send the reward to the reporter.
+		code := rt.Send(reporter, builtin.MethodSend, nil, toReward, &builtin.Discard{})
+
+		// If we fail, log and burn the reward to make sure the balances remain correct.
+		if !code.IsSuccess() {
+			rt.Log(rtt.ERROR, "failed to send reward")
+			toBurn = big.Add(toBurn, toReward)
```

Comment:
This "burn it" isn't present in consensus fault reporting, but it's necessary to ensure that pledge values remain consistent.

Alternatively, we _could_ just fail the message. Honestly, _not_ failing the message is a bit of a footgun as lotus will think "the message succeeded" and will just send it to the network instead of kicking it back to the user to ensure the user gets paid.

Reply:
I'd rather the report succeeded even if we can't pay the reward. Burning it sounds good.

Reply:
For something like this, I assume the submitter would just try again. Any reason we need it to succeed the first time?

Reply:
Why would it succeed the second time? I don't know what might have gone wrong to cause this failure (low balance?) but it is likely to be permanent. I don't think we should allow inability to pay the reward to stop the fraud being reported at all.

Reply:
Hm. I was thinking something like the actor that sent the message suddenly got deleted (not sure how this can happen...).

But this isn't going to fix the "low balance" issue. If the actor doesn't have enough funds for the reward, they'll fail on the next step (burning).